### PR TITLE
Scheduler changes to introduce alpha support for Pod Overhead

### PR DESCRIPTION
**Pull request by egernst**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

This PR adds initial support for alpha-feature PodOverhead as part of the changes described in the  [Pod Overhead KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190226-pod-overhead.md).

**Special notes for your reviewer**:

This just handles requests, and as is doesn't provide great error messages.  May want to do further refactoring, separating explicit requests from the pod overhead in order to provide better error messages.
 
```release-note
Introduction of the pod overhead feature to the scheduler.  This functionality is alpha-level as of 
Kubernetes v1.16, and is only honored by servers that enable the PodOverhead feature.gate. 
```

### CODE REVIEW

**Comment by egernst** [comment]
PTAL at initial changes in just this commit: https://github.com/kubernetes/kubernetes/pull/78319/commits/faed38c8d9edcf83535d149c87186b65ba0b43cd

API changes are handled in https://github.com/kubernetes/kubernetes/pull/76968

File: pkg/scheduler/algorithm/predicates/predicates_test.go
Comment (ahg-g) [inline pkg/scheduler/algorithm/predicates/predicates_test.go]: why those tests were removed?

File: pkg/scheduler/nodeinfo/node_info.go
Comment (ahg-g) [inline pkg/scheduler/nodeinfo/node_info.go]: ditto regarding minimum values returned by GetNozeroRequests.

File: pkg/scheduler/algorithm/priorities/resource_allocation.go
Comment (ahg-g) [inline pkg/scheduler/algorithm/priorities/resource_allocation.go]: This means if cpu/memory are not set in pod.Spec.Overhead, a default value of .1 vcpu/200MB will be returned. Is it the intention that those are the minimum overheads?

File: pkg/apis/node/v1alpha1/conversion.go
Comment (egernst) [reply pkg/apis/node/v1alpha1/conversion.go]: https://github.com/egernst/kubernetes/blob/scheduler-changes/pkg/apis/node/v1alpha1/zz_generated.conversion.go#L89-L92

File: pkg/scheduler/algorithm/predicates/predicates_test.go
Comment (egernst) [reply pkg/scheduler/algorithm/predicates/predicates_test.go]: error on my side -- addressed.

File: pkg/scheduler/algorithm/priorities/resource_allocation.go
Comment (egernst) [reply pkg/scheduler/algorithm/priorities/resource_allocation.go]: Got it - thanks @ahg-g - this is definitely not the intention.  TBH, how  nonzero requests are used (and how they are related to the total requests) was not very clear to me when following the code/definitions.

It *sounds* like GetNonZero requests wouldn't need Overhead then?  How's `non0CPUReq, non0MemReq` used?

The intent is to take into account the overhead associated with running a sandbox (pod), and from an "allocatable" standpoint to take this into account.

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (egernst) [reply pkg/scheduler/algorithm/predicates/predicates.go]: done.

File: pkg/scheduler/algorithm/priorities/resource_allocation.go
Comment (ahg-g) [reply pkg/scheduler/algorithm/priorities/resource_allocation.go]: As we discussed over slack, we should have a separate function that basically return 0 memory and cpu if their values were not set.

Comment (egernst) [reply pkg/scheduler/algorithm/priorities/resource_allocation.go]: Thanks for the help @ahg-g.  updated.

File: pkg/api/pod/util.go
Comment (bsalamat) [inline pkg/api/pod/util.go]: nit: remove the blank line please

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (bsalamat) [inline pkg/scheduler/algorithm/predicates/predicates.go]: Checking feature gates in the critical path of scheduling has shown to cause performance degradation. Feature gates should be read once and passed to the scheduler.

File: pkg/scheduler/algorithm/priorities/resource_allocation.go
Comment (bsalamat) [inline pkg/scheduler/algorithm/priorities/resource_allocation.go]: comments should start with the name of function, struct, etc. Please start the comment with `getNonZeroRequests`.

File: pkg/scheduler/algorithm/priorities/resource_allocation.go
Comment (bsalamat) [inline pkg/scheduler/algorithm/priorities/resource_allocation.go]: Similar comment about checking feature gates applies here.

File: pkg/scheduler/nodeinfo/node_info.go
Comment (bsalamat) [inline pkg/scheduler/nodeinfo/node_info.go]: ditto

**Review by bsalamat — COMMENTED** [review]
Thanks, @egernst! The scheduler changes look generally good. I have a few minor comments.

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (egernst) [reply pkg/scheduler/algorithm/predicates/predicates.go]: Understood.  I am looking now for a pre-existing example, though nothing immediate comes up (looking at commit history and scheduler pkg).

File: pkg/apis/core/validation/validation.go
Comment (draveness) [inline pkg/apis/core/validation/validation.go]: small nit: extra space after `//`

File: pkg/apis/core/validation/validation.go
Comment (draveness) [inline pkg/apis/core/validation/validation.go]: It would be better to remove the resourceReq with the following code which would make the code more concise.

```go
	return ValidateResourceRequirements(&core.ResourceRequirements{
		Limits: overhead,
	}, fldPath)

```

File: pkg/apis/core/validation/validation_test.go
Comment (draveness) [inline pkg/apis/core/validation/validation_test.go]: nit: extra line here

File: pkg/apis/node/validation/validation.go
Comment (draveness) [inline pkg/apis/node/validation/validation.go]: same as [1]

File: pkg/apis/core/validation/validation.go
Comment (draveness) [reply pkg/apis/core/validation/validation.go]: [1]

File: pkg/scheduler/nodeinfo/node_info_test.go
Comment (draveness) [inline pkg/scheduler/nodeinfo/node_info_test.go]: nit: extra line

File: pkg/scheduler/nodeinfo/node_info_test.go
Comment (draveness) [inline pkg/scheduler/nodeinfo/node_info_test.go]: nit: extra line

File: pkg/apis/node/validation/validation_test.go
Comment (draveness) [inline pkg/apis/node/validation/validation_test.go]: nit: extra new line

**Review by draveness — CHANGES_REQUESTED** [review]

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (bsalamat) [reply pkg/scheduler/algorithm/predicates/predicates.go]: Here is an example: https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/factory/factory.go#L290

Comment (egernst) [reply pkg/scheduler/algorithm/predicates/predicates.go]: Thanks @bsalamat.  In the case of enableNonPreempting, it is relatively straight forward since the consumption of the flag is limited to within the genericScheduler itself.

In the case of enablePodOverhead, it is needed in three places (at least):
 - [node-info](https://github.com/egernst/kubernetes/blob/scheduler-changes/pkg/scheduler/nodeinfo/node_info.go#L591-L602): 
```golang
	// If Overhead is being utilized, add to the total requests for the pod
	if pod.Spec.Overhead != nil && utilfeature.DefaultFeatureGate.Enabled(features.PodOverhead) {
		resPtr.Add(pod.Spec.Overhead)

		if _, found := pod.Spec.Overhead[v1.ResourceCPU]; found {
			non0CPU += pod.Spec.Overhead.Cpu().MilliValue()
		}

		if _, found := pod.Spec.Overhead[v1.ResourceMemory]; found {
			non0Mem += pod.Spec.Overhead.Memory().Value()
		}
	}
```
 - [predicate's GetResourceRequest](https://github.com/egernst/kubernetes/blob/scheduler-changes/pkg/scheduler/algorithm/predicates/predicates.go#L746-L763)
 - [priority's resource allocation requests](https://github.com/egernst/kubernetes/blob/scheduler-changes/pkg/scheduler/algorithm/priorities/resource_allocation.go#L94-L116)

The problem I have is finding the appropriate structure to stash the PodOverhead flag in.

## First attempt -- adding to NodeInfo structure:
NodeInfo is available for each area, but this doesn't work well, since: 1) feature gate is looked up per node, and 2) the GetMetadata for both priority and [predicates](https://github.com/egernst/kubernetes/blob/scheduler-changes/pkg/scheduler/algorithm/predicates/metadata.go#L155) calls into the applicable function where node specific information isn't available (ie, GetMetadata operates on a set of nodes)

## Second attempt -- add to the {Priority|Predicate}MetadataFactory

This isn't straightforward either, but may be next 'best' option. Modifications would be needed within the algorithmprovider for creating New{Priority|Predicate}MetadataFactory.

WDYT @bsalamat ? This match your expectation?

File: pkg/scheduler/algorithm/priorities/resource_allocation.go
Comment (tallclair) [inline pkg/scheduler/algorithm/priorities/resource_allocation.go]: What is the effect of adding the overhead here if the container requests were explicitly set to zero?

File: pkg/scheduler/algorithm/predicates/predicates.go
Comment (bsalamat) [reply pkg/scheduler/algorithm/predicates/predicates.go]: Thanks for checking the options. You are right that this particular feature gate is harder to pass to various parts of the scheduler. Unfortunately, none of the possible options are sufficiently clean. So, let's keep them as they are. Since we check  `pod.Spec.Overhead != nil` before checking the feature gate, this will not have a performance impact for pods that don't use the feature.

File: pkg/scheduler/algorithm/priorities/resource_allocation.go
Comment (egernst) [reply pkg/scheduler/algorithm/priorities/resource_allocation.go]: Interesting - I'm curious the scenario where this would happen.

Either way, if they are running their explicitly-set-zero-request workload in a sandbox (RuntimeClass with Overhead), there is still a cost of running that sandbox, so the result of this call will be just the pod overhead.

To me, this makes sense.  WDYT?

Comment (tallclair) [reply pkg/scheduler/algorithm/priorities/resource_allocation.go]: Yeah, makes sense, but I think I don't really understand this method. I think maybe it should just be called "getRequestsSum" - I think non-zero just means apply defaulting?

Comment (egernst) [reply pkg/scheduler/algorithm/priorities/resource_allocation.go]: Correct -- for accounting purposes it adds a non-zero value for requests that were zero.  The naming and utilization wasn't very clear to me at first.  In either case, though, the Overhead addition is needed.  There was another PR opened to help address the naming/usage: see https://github.com/kubernetes/kubernetes/pull/78263

Comment (egernst) [reply pkg/scheduler/algorithm/priorities/resource_allocation.go]: - `requestedResources` is the exact sum of pod resource requests on that node.
- `nonzeroRequest` is a modified version where we apply a minimum value to CPU and memory for any container that doesn't specify a value.

**Comment by ahg-g** [comment]
you may want to run ./hack/update-bazel.sh to address the failed build test.

**Comment by ahg-g** [comment]
/lgtm

Can you please open an issue to add a scheduler integration test for this feature? unless you want to add it in this PR :)

**Review by bsalamat — COMMENTED** [review]
/approve

Thanks, @egernst!

**Comment by egernst** [comment]
@ahg-g -- I opened an issue: see https://github.com/kubernetes/kubernetes/issues/80264

I had to rebase due to a trivial conflict.  This PR needs a new 'LGTM,' if you're willing.

**Review by draveness — APPROVED** [review]
/retest
/priority important-soon

**Pull request merged**
